generic_config_updater/gu_common.py

```diff
@@ -174,3 +188,562 @@ def convert_sonic_yang_patch_to_config_db_patch(self, patch):
+                    "Ethernet8": {}
+                }
+            }
+          return:
+            /VLAN_MEMBER/Vlan1000|Ethernet0
+            /VLAN_MEMBER/Vlan1000|Ethernet4
+            /ACL_TABLE/EVERFLOW/ports/0
+            /ACL_TABLE/EVERFLOW6/ports/0
+            /ACL_TABLE/EVERFLOW6/ports/1
+        """
+        # TODO: Also fetch references by must statement (check similar statements)
+        return self._find_leafref_paths(path, config)
+
+    def _find_leafref_paths(self, path, config):
+        sy = sonic_yang.SonicYang(YANG_DIR)
+        sy.loadYangModel()
+
+        sy.loadData(config)
+
+        xpath = self.convert_path_to_xpath(path, config, sy)
+
+        leaf_xpaths = self._get_inner_leaf_xpaths(xpath, sy)
+
+        ref_xpaths = []
+        for xpath in leaf_xpaths:
+            ref_xpaths.extend(self._find_leaf_dependencies(sy, xpath))
+
+        ref_paths = []
+        for ref_xpath in ref_xpaths:
+            ref_path = self.convert_xpath_to_path(ref_xpath, config, sy)
+            ref_paths.append(ref_path)
+
+        return set(ref_paths)
+
+    def _get_inner_leaf_xpaths(self, xpath, sy):
+        if xpath=="/": # Point to Root element which contains all xpaths
+            nodes = sy.root.tree_for()
+        else: # Otherwise get all nodes that match xpath
+            nodes = sy.root.find_path(xpath).data()
+
+        for node in nodes:
+            for inner_node in node.tree_dfs():
+                # TODO: leaflist also can be used as the 'path' argument in 'leafref' so add support to leaflist
+                if self._is_leaf_node(inner_node):
+                    yield inner_node.path()
+
+    def _is_leaf_node(self, node):
+        schema = node.schema()
+        return ly.LYS_LEAF == schema.nodetype() # or ly.LYS_LEAFLIST == schema.nodetype()
+
+    # TODO: YANG SONiC lib has this function but with a bug in handling null backlinks.
+    #       It is OK to have null backlinks if there is no backlinks.
+    def _find_leaf_dependencies(self, sy, data_xpath):
+        ref_list = []
+        node = sy.root
+        try:
+            data_node = sy._find_data_node(data_xpath)
+        except Exception as e:
+            print("find_data_dependencies(): Failed to find data node from xpath: {}".format(data_xapth))
+            return ref_list
+
+        try:
+            value = str(sy._find_data_node_value(data_xpath))
+
+            schema_node = ly.Schema_Node_Leaf(data_node.schema())
+            backlinks = schema_node.backlinks()
+            if backlinks is not None and backlinks.number() > 0:
+                for link in backlinks.schema():
+                     node_set = node.find_path(link.path())
+                     for data_set in node_set.data():
+                          data_set.schema()
+                          casted = data_set.subtype()
+                          if value == casted.value_str():
+                              ref_list.append(data_set.path())
+        except Exception as e:
```

Comment:
>Exception [](http://example.com/codeflow?start=15&length=9)

Could you use specific exception type? Seems you know the exception is related to "Failed to find node or dependencies"

Reply:
code removed

Reply:
code removed